### PR TITLE
Propagate DataContext from ToolTip to Popup

### LINKED ISSUE #9570: [WASM] Binding in TooltipService.Tooltip.Content
### Current behavior

```TooltipService.Content``` can be set to a literal, however, binding it to a property doesn't work.
see: https://github.com/paulovila/TooltipService 
image

### Expected behavior

_No response_

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

_No response_

### Environment

_No response_

### NuGet package version(s)

4.5.0-dev.562

### Affected platforms

WebAssembly

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #9570

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ToolTipTests/ToolTip_Tests.cs [hunk 1]
```diff
@@ -120,5 +120,31 @@ public void ToolTip_WhenHostCollapsed_Test()
...
+			setDataContextValueTextBlockButton.FastTap();
+			Assert.AreEqual("1", dataContextValueTextBlock.GetDependencyPropertyValue<string>("Text"));
+
+			updateDataContextButton.FastTap();
+			setDataContextValueTextBlockButton.FastTap();
+			// TODO: This is a bug. It should be 2
+			Assert.AreEqual("1", dataContextValueTextBlock.GetDependencyPropertyValue<string>("Text"));
+			setDataContextValueTextBlockButton.FastTap();
+			// TODO: This is a bug. It should be 2
+			Assert.AreEqual("1", dataContextValueTextBlock.GetDependencyPropertyValue<string>("Text"));
```
Comment (Youssef1313) [inline src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ToolTipTests/ToolTip_Tests.cs:147]: I spent too much trying to get this scenario to work. But I always break something.

Comment (Youssef1313) [reply src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ToolTipTests/ToolTip_Tests.cs:147]: I'll open a new issue for this if the PR got merged.

File: src/Uno.UI/UI/Xaml/Controls/ToolTip/ToolTip.cs [hunk 2]
```diff
@@ -40,6 +40,7 @@ internal Popup Popup
 					_popup.PopupPanel = new PopupPanel(_popup);
 					_popup.Opened += OnPopupOpened;
 					_popup.Closed += (sender, e) => IsOpen = false;
+					_popup.DataContext = DataContext;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/ToolTip/ToolTip.cs:43]: Did you try adding what I mentioned here: https://github.com/unoplatform/uno/issues/9570#issuecomment-1227190049 ?

File: src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ToolTipTests/ToolTip_Tests.cs (hunk 1 above)
Comment (jeromelaban) [reply src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ToolTipTests/ToolTip_Tests.cs:147]: It's probably best if create a runtime test for that instead (and it'll run on all platforms). You can use internal APIs that tap a button and then test using `FindName`.

File: src/Uno.UI/UI/Xaml/Controls/ToolTip/ToolTip.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/ToolTip/ToolTip.cs:43]: @jeromelaban Yeah, OnDataContextChanged isn't called.

The reason is when it gets changed, the precedence is `Inheritance`, but it already has a `Local` precedence.

File: src/Uno.UI/UI/Xaml/Controls/Popup/Popup.Base.cs [hunk 3]
```diff
@@ -119,7 +119,11 @@ protected override Size ArrangeOverride(Size finalSize)
 				provider.Store.ClearValue(AllowFocusWhenDisabledProperty, DependencyPropertyValuePrecedences.Local);
 			}
 
-			UpdateDataContext(null);
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Popup/Popup.Base.cs:122]: @ajpinedam would you remember why you had to do this special case?

File: src/Uno.UI/UI/Xaml/Controls/ToolTip/ToolTip.cs [hunk 4]
```diff
@@ -141,7 +141,16 @@ private void AttachToPopup()
...
+				var previousParent = this.GetParent();
+
 				Popup.Child = this;
+
+				// The tooltip is integrated into the PopupPanel, which automatically
+				// propagates its own datacontext through inheritance. This ends up changing
+				// the DataContext of the tooltip to a local value which cannot be overriden
+				// through the tooltip's own actual parent (the control that owns the TooltipService).
+				// Restoring the original parent ensures that an incorrect DataContext property will not flow through.
+				this.SetParent(previousParent);
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/ToolTip/ToolTip.cs:153]: Very nice!!

Thanks @jeromelaban

File: src/Uno.UI/UI/Xaml/Controls/Popup/Popup.Base.cs [hunk 5]
```diff
@@ -157,7 +165,7 @@ private void UpdateDataContext(DependencyPropertyChangedEventArgs e)
 					// Child already has locally set DataContext, we shouldn't overwrite it.
 					_childHasOwnDataContext = true;
 				}
-				else
+				else if(PropagatesDataContextToChild)
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/Popup/Popup.Base.cs:168]: Is there a problem if we skipped the whole `UpdateDataContext` on this condition?

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/Popup/Popup.Base.cs]: There isn't, let's do that for now :)

**Review by jeromelaban — APPROVED** [review]

**Review by ajpinedam — APPROVED** [review]
Nice! Looks good.

**Pull request merged**